source/mir/sparse/package.d

```diff
@@ -0,0 +1,614 @@
+         [6, 0, 9],
+         [0, 0, 5]];
+
+    auto crs = sparse.compress;
+    assert(crs.ptr.range == CompressedMap!(double, uint, uint)(
+         3,
+        [2, 1, 4, 6, 9, 5],
+        [1, 2, 2, 0, 2, 2],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Sparse tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
+
+    auto sparse = sparse!double(5, 8);
+    sparse[] =
+        [[0, 2, 0, 0, 0, 0, 0, 1],
+         [0, 0, 0, 0, 0, 0, 0, 4],
+         [0, 0, 0, 0, 0, 0, 0, 0],
+         [6, 0, 0, 0, 0, 0, 0, 9],
+         [0, 0, 0, 0, 0, 0, 0, 5]];
+
+    auto crs = sparse.compress;
+    assert(crs.ptr.range == CompressedMap!(double, uint, uint)(
+         8,
+        [2, 1, 4, 6, 9, 5],
+        [1, 7, 7, 0, 7, 7],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Dense tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
+
+    auto slice = slice!double(5, 3);
+    slice[] =
+        [[0, 2, 1],
+         [0, 0, 4],
+         [0, 0, 0],
+         [6, 0, 9],
+         [0, 0, 5]];
+
+    auto crs = slice.compress;
+
+    assert(crs.ptr.range == CompressedMap!(double, uint, uint)(
+         3,
+        [2, 1, 4, 6, 9, 5],
+        [1, 2, 2, 0, 2, 2],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Dense tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
+
+    auto slice = slice!double(5, 8);
+    slice[] =
+        [[0, 2, 0, 0, 0, 0, 0, 1],
+         [0, 0, 0, 0, 0, 0, 0, 4],
+         [0, 0, 0, 0, 0, 0, 0, 0],
+         [6, 0, 0, 0, 0, 0, 0, 9],
+         [0, 0, 0, 0, 0, 0, 0, 5]];
+
+    auto crs = slice.compress;
+    assert(crs.ptr.range == CompressedMap!(double, uint, uint)(
+         8,
+        [2, 1, 4, 6, 9, 5],
+        [1, 7, 7, 0, 7, 7],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+///
+CompressedTensor!(N, V, I, J)
+    compressWithType
+    (V, I = uint, J = uint, S : Slice!(N, R), size_t N, R : SparseMap!T, T)
+    (S slice)
+    if(is(T : V) && N > 1)
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    import mir.ndslice.selection: iotaSlice;
+    auto data = slice
+        .ptr
+        .range
+        .table
+        .byKeyValue
+        .array
+        .sort!((a, b) => a.key < b.key)
+        .release;
+    auto inv = iotaSlice(slice.shape[0 .. N - 1]);
+    auto count = inv.elementsCount;
+    auto map = CompressedMap!(V, I, J)(
+        slice.length!(N - 1),
+        new V[data.length],
+        new I[data.length],
+        new J[count + 1],
+        );
+    size_t k = 0;
+    map.pointers[0] = 0;
+    map.pointers[1] = 0;
+    foreach(t, e; data)
+    {
+        map.values[t] = cast(V) e.value;
+        size_t index = e.key;
+        map.indexes[t] = cast(I)(index % slice.length!(N - 1));
+        auto p = index / slice.length!(N - 1);
```

Comment:
maybe we can store the result of `slice.length!(N - 1) in one variable? would cleanup the loop.